plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -246,64 +245,69 @@ func findNodesThatFit(
 	metadataProducer algorithm.PredicateMetadataProducer,
 	ecache *EquivalenceCache,
 ) ([]*v1.Node, FailedPredicateMap, error) {
-	var filtered []*v1.Node
+
 	failedPredicateMap := FailedPredicateMap{}
+	filtered := nodes
 
-	if len(predicateFuncs) == 0 {
-		filtered = nodes
-	} else {
+	if len(predicateFuncs) > 0 {
 		// Create filtered list with enough space to avoid growing it
 		// and allow assigning.
-		filtered = make([]*v1.Node, len(nodes))
-		errs := errors.MessageCountMap{}
-		var predicateResultLock sync.Mutex
-		var filteredLen int32
+		filtered = make([]*v1.Node, 0, len(nodes))
 
 		// We can use the same metadata producer for all nodes.
 		meta := metadataProducer(pod, nodeNameToInfo)
+
+		errs := errors.MessageCountMap{}
+		var predicateResultLock sync.Mutex
+
 		checkNode := func(i int) {
 			nodeName := nodes[i].Name
+
 			fits, failedPredicates, err := podFitsOnNode(pod, meta, nodeNameToInfo[nodeName], predicateFuncs, ecache)
 			if err != nil {
 				predicateResultLock.Lock()
 				errs[err.Error()]++
 				predicateResultLock.Unlock()
 				return
 			}
+
 			if fits {
-				filtered[atomic.AddInt32(&filteredLen, 1)-1] = nodes[i]
+				filtered = append(filtered, nodes[i])
```

Comment:
Now there is a race condition in appending to `filtered`. Previous code was written with atomic add to remove the race condition.

Reply:
Didn't notice that.. I could put the lock around it. Good catch.

Reply:
No. Locks are slow. The previous version of the code used an atomic variable to achieve lock-free append. Please keep it as it was.